cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -1617,7 +1618,13 @@ private module ExprFlowCached {
    */
   private predicate localStepFromNonExpr(Node n1, Node n2) {
     not exists(n1.asExpr()) and
-    localFlowStep(n1, n2)
+    localFlowStep(n1, n2) and
+    // We should never recursive through any of these nodes while
+    // looking for the next expression.
+    not n1 instanceof SsaPhiNode and
+    not n1 instanceof InitialGlobalValue and
```

Comment:
The comment doesn't quite give me sufficient information here as to why this is ok.

Reply:
Hm, yes I can see what you mean. I'll expand the explanation. Taking a step back, the point of `localStepsToExpr(n1, n2, e2)` is that `n2` is the first node after `n1` in the dataflow step relation that has a result for `asExpr()`. So `localStepsToExpr` does a transitive closure over `localFlowStep`, but where `localFlowStep` is restricted to only take a step as long as there's no result for `asExpr()` (this is the `localStepFromNonExpr` predicate). Once we find a node for which `asExpr()` exists, the transitive closure stops.

On the particular project I was looking at, there apparently were some nasty `SsaPhiNode`s that made the transitive closure of `localStepFromNonExpr` blow up even with the restriction mentioned above. `SsaPhiNode` are trivially includes in `localStepFromNonExpr` because both `SsaPhiNode` because they clearly satisfy `not exists(n1.asExpr())` (as these nodes are not supposed to map to expressions) and they can have many many incoming nodes. This resulted in the tuple explosion.

But when we're looking for the next node after `n1` in the dataflow step relation that has a result for `asExpr()`, we should never step over an `SsaPhiNode` since this node represents incoming data from many different places.

I'll try to write a more carefully worded description of the above and put it in a comment if it makes more sense now?

Reply:
https://github.com/github/codeql/pull/12477/commits/37d417bb4550479c60f1aec1862316c87f0aed74 adds a version of ^ into the code.

Reply:
This makes sense now.